File: gunhoflash/maximum-product-subarray.ts
function maxProduct(nums: number[]): number {
  let answer = Math.max(...nums);

  let minPositiveSerialProduct = 1;
  let maxNegativeSerialProduct = 0;
  let serialProduct = 1;

  for (const num of nums) {
    serialProduct *= num;

    if (serialProduct < 0) {
      if (maxNegativeSerialProduct === 0) {
        maxNegativeSerialProduct = serialProduct;
      } else {
        answer = Math.max(answer, serialProduct / maxNegativeSerialProduct);
        maxNegativeSerialProduct = Math.max(maxNegativeSerialProduct, serialProduct);
      }
    } else if (serialProduct > 0) {
      answer = Math.max(answer, serialProduct / minPositiveSerialProduct);
      minPositiveSerialProduct = Math.min(minPositiveSerialProduct, serialProduct);
    } else {
      minPositiveSerialProduct = 1;
      maxNegativeSerialProduct = 0;
      serialProduct = 1;
    }
  }

  return answer;
}

File: gunhoflash/maximum-product-subarray.py
from typing import List

class Solution:
  def maxProduct(self, nums: List[int]) -> int:
    answer = max(nums)

    minPositiveSerialProduct = 1
    maxNegativeSerialProduct = 0
    serialProduct = 1

    for num in nums:
      serialProduct *= num

      if serialProduct < 0:
        if maxNegativeSerialProduct == 0:
          maxNegativeSerialProduct = serialProduct
        else:
          answer = max(answer, serialProduct // maxNegativeSerialProduct)
          maxNegativeSerialProduct = max(maxNegativeSerialProduct, serialProduct)
      elif serialProduct > 0:
        answer = max(answer, serialProduct // minPositiveSerialProduct)
        minPositiveSerialProduct = min(minPositiveSerialProduct, serialProduct)
      else:
        minPositiveSerialProduct = 1
        maxNegativeSerialProduct = 0
        serialProduct = 1

    return answer
